_source/_posts/2021-12-14-antipatterns-secrets-java.adoc

```diff
@@ -0,0 +1,147 @@
+---
+layout: blog_post
+title: "Five Anti-Patterns with Secrets in Java"
+author: brian-demers
+by: advocate
+communities: [security,java]
+description: "Learn how how to better manage and protect your passwords and API keys!"
+tags: [java, security, secrets, cryptography, tips]
+tweets:
+- "When was the last time you checked a password into source control? 🤦 Here are a few tips to avoid that and more‼️"
+- "API keys, passwords, and secrets. Learn a few common anti-patterns to avoid ⚠️"
+- "Secrets, passwords, and API keys oh my‼️ Make sure you are handling them correctly 🧑‍💻"
+image:
+type: awareness
+---
+
+:toc: macro
+:page-liquid:
+:experimental:
+
+Most applications require some sort of secret or password to enable access: database connection info, API keys, OAuth client secrets, and JWT keys, to list some examples. Dealing with secrets in your projects is always a chore, and it's often done wrong. In this post, I'll describe five common problems, which you can think of as *anti-patterns* (the opposite of a *best practice*) and I'll offer suggestions to help you avoid these issues.
+
+toc::[]
+
+== 1. Don't check your secrets into source control
+
+Checking a secret into source control is a mistake that's easily made. This is an anti-pattern that most of us have done in error at some point, even if we don't want to admit it. Once a secret is in your source control, it's there forever; it only takes https://github.com/search?q=removed+password&type=commits[a quick history search] to find.
+
+If you find a secret in your source, remove it and revoke it. Of course, it's best if secrets are never checked in to begin with! Watch out for them in code reviews and scan your code bases to ensure this doesn't happen: https://www.gitguardian.com/[GitGuardian] and https://github.com/awslabs/git-secrets[git-secrets].
+
+Secrets often enter source trees by accident, usually because you temporarily typed one into a config file. When secrets are mixed in with other application properties, they become harder to spot and easier to forget about. Keep secrets out of your source entirely by using an https://12factor.net/config[external configuration] provider like [Hashicorp Vault](https://www.vaultproject.io/[Hashicorp Vault]), using environmental variables, or by keeping secrets in a separate file outside of your source tree (or one that is explicitly ignored in your `.gitignore` file.)
```

Comment:
```suggestion
Secrets often enter source trees by accident, usually because you temporarily typed one into a config file. When secrets are mixed in with other application properties, they become harder to spot and easier to forget about. Keep secrets out of your source entirely by using an https://12factor.net/config[external configuration] provider like https://www.vaultproject.io/[HashiCorp Vault], using environmental variables, or by keeping secrets in a separate file outside of your source tree (or one that is explicitly ignored in your `.gitignore` file.)
```